File: src/main/java/com/rdvdev2/timetravelmod/api/timemachine/TimeMachineBuilder.java
package com.rdvdev2.timetravelmod.api.timemachine;

import com.rdvdev2.timetravelmod.impl.common.timemachine.TimeMachine;
import net.minecraft.block.BlockState;
import net.minecraft.item.Item;

public class TimeMachineBuilder {

    private Item icon = null;
    private int cooldownTime = 400;
    private int tier = 1;
    private String[][] structureLayers = null;
    private BlockState[] controllerStates = null;
    private BlockState[] coreStates = null;
    private BlockState[] basicStates = null;
    private int entityMaxLoad = 1;
    private int corruptionMultiplier = 1;

    public TimeMachineBuilder setIcon(Item icon) {
        this.icon = icon;
        return this;
    }

    public TimeMachineBuilder setCooldownTime(int cooldownTime) {
        this.cooldownTime = cooldownTime;
        return this;
    }

    public TimeMachineBuilder setTier(int tier) {
        this.tier = tier;
        return this;
    }

    public TimeMachineBuilder setStructureLayers(String[][] structureLayers) {
        this.structureLayers = structureLayers;
        return this;
    }

    public TimeMachineBuilder setControllerStates(BlockState... controllerStates) {
        this.controllerStates = controllerStates;
        return this;
    }

    public TimeMachineBuilder setCoreStates(BlockState... coreStates) {
        this.coreStates = coreStates;
        return this;
    }

    public TimeMachineBuilder setBasicStates(BlockState... basicStates) {
        this.basicStates = basicStates;
        return this;
    }

    public TimeMachineBuilder setEntityMaxLoad(int entityMaxLoad) {
        this.entityMaxLoad = entityMaxLoad;
        return this;
    }

    public TimeMachineBuilder setCorruptionMultiplier(int corruptionMultiplier) {
        this.corruptionMultiplier = corruptionMultiplier;
        return this;
    }

    public ITimeMachine build() {
        if(structureLayers == null || controllerStates == null || coreStates == null || basicStates == null)
            throw new RuntimeException("You must define all properties");
        return new TimeMachine(icon, cooldownTime, tier, structureLayers, controllerStates, coreStates, basicStates, entityMaxLoad, corruptionMultiplier);
    }
}

File: src/main/kotlin/com/rdvdev2/timetravelmod/api/timemachine/TimeMachineBuilder.kt
package com.rdvdev2.timetravelmod.api.timemachine

import com.rdvdev2.timetravelmod.impl.common.timemachine.TimeMachine
import net.minecraft.block.BlockState
import net.minecraft.item.Item

class TimeMachineBuilder {

    private var icon: Item? = null
    private var cooldownTime = 400
    private var tier = 1
    private var structureLayers: Array<Array<String>> = emptyArray()
    private var controllerStates: Array<BlockState> = emptyArray()
    private var coreStates: Array<BlockState> = emptyArray()
    private var basicStates: Array<BlockState> = emptyArray()
    private var entityMaxLoad = 1
    private var corruptionMultiplier = 1

    fun setIcon(icon: Item?): TimeMachineBuilder {
        this.icon = icon
        return this
    }

    fun setCooldownTime(cooldownTime: Int): TimeMachineBuilder {
        this.cooldownTime = cooldownTime
        return this
    }

    fun setTier(tier: Int): TimeMachineBuilder {
        this.tier = tier
        return this
    }

    fun setStructureLayers(structureLayers: Array<Array<String>>): TimeMachineBuilder {
        this.structureLayers = structureLayers
        return this
    }

    fun setControllerStates(vararg controllerStates: BlockState): TimeMachineBuilder {
        this.controllerStates = arrayOf(*controllerStates)
        return this
    }

    fun setCoreStates(vararg coreStates: BlockState): TimeMachineBuilder {
        this.coreStates = arrayOf(*coreStates)
        return this
    }

    fun setBasicStates(vararg basicStates: BlockState): TimeMachineBuilder {
        this.basicStates = arrayOf(*basicStates)
        return this
    }

    fun setEntityMaxLoad(entityMaxLoad: Int): TimeMachineBuilder {
        this.entityMaxLoad = entityMaxLoad
        return this
    }

    fun setCorruptionMultiplier(corruptionMultiplier: Int): TimeMachineBuilder {
        this.corruptionMultiplier = corruptionMultiplier
        return this
    }

    fun build(): ITimeMachine {
        return TimeMachine(icon, cooldownTime, tier, structureLayers, controllerStates, coreStates, basicStates, entityMaxLoad, corruptionMultiplier)
    }
}